src/sc/sidechain.cpp

```diff
@@ -247,68 +247,104 @@ bool ScMgr::hasScCreationConflictsInMempool(const CTxMemPool& pool, const CTrans
+{
     if (initDone)
     {
-        LogPrintf("%s():%d - Error: could not init from db mpre than once!\n", __func__, __LINE__);
-        return false;
+    	return error("%s():%d - could not init from db more than once!", __func__, __LINE__);
     }
 
-    db = new CLevelDBWrapper(GetDataDir() / "sidechains", cacheSize, false, fWipe);
 
     initDone = true;
+    chosenDbCreationPolicy = dbPolicy;
 
-    LOCK(sc_lock);
-
-    boost::scoped_ptr<leveldb::Iterator> it(db->NewIterator());
+    if (dbPolicy == dbCreationPolicy::mock)
+    {
+    	return true; //db is not instantiated and mScInfo is kept initially empty
+    }
 
-    for (it->SeekToFirst(); it->Valid(); it->Next())
+    if (dbPolicy == dbCreationPolicy::create)
     {
-        boost::this_thread::interruption_point();
+    	db = new CLevelDBWrapper(GetDataDir() / "sidechains", cacheSize, false, fWipe);
 
-        try {
-            leveldb::Slice slKey = it->key();
-            CDataStream ssKey(slKey.data(), slKey.data()+slKey.size(), SER_DISK, CLIENT_VERSION);
-            char chType;
-            uint256 keyScId;
-            ssKey >> chType;
-            ssKey >> keyScId;;
- 
-            if (chType == DB_SC_INFO)
-            {
-                leveldb::Slice slValue = it->value();
-                CDataStream ssValue(slValue.data(), slValue.data()+slValue.size(), SER_DISK, CLIENT_VERSION);
-                ScInfo info;
-                ssValue >> info;
- 
-                mScInfo[keyScId] = info;
-                LogPrint("sc", "%s():%d - scId[%s] added in map\n", __func__, __LINE__, keyScId.ToString() );
-            }
-            else
-            {
-                // should never happen
-                LogPrintf("%s():%d - Error: could not read from db, invalid record type %c\n", __func__, __LINE__, chType);
-                return false;
-            }
-        }
+    	//load initial data!
+    	LOCK(sc_lock);
+    	try
+    	{
+    		bool res = loadInitialDataFromDb();
+    		if (!res)
+    		{
+    			return error("%s():%d - error occurred during db scan", __func__, __LINE__);
+    		}
+    	}
         catch (const std::exception& e)
         {
             return error("%s: Deserialize or I/O error - %s", __func__, e.what());
         }
-    }
 
-    if (!it->status().ok())
-    {
-        return error("%s():%d - error occurred during db scan", __func__, __LINE__);
+        return true;
     }
-    return true;
+
+    return error("%s():%d - error specifying db creation policy", __func__, __LINE__);
+}
+
+void ScMgr::reset()
+{
+	delete db;
+	db = nullptr;
+	mScInfo.clear();
+	initDone = false;
+	chosenDbCreationPolicy = dbCreationPolicy::create; //the original one
 }
 
 void ScMgr::eraseFromDb(const uint256& scId)
 {
+	if (chosenDbCreationPolicy == dbCreationPolicy::mock)
+	{
+		return; //nothing to erase from db
+	}
+
+	if (chosenDbCreationPolicy != dbCreationPolicy::create)
+	{
+		error("%s():%d - error specifying db creation policy", __func__, __LINE__);
```

Comment:
similarly to the code used elsewhere:
return error("<msg>");

Reply:
agree. change introduced